### ISSUE TITLE
Change applied to Blazor WebAssembly app using .NET Hot Reload don't show up after adding binding code

### ISSUE THREAD

**Issue by danroth27** [dotnet/aspnetcore#34446]
Repro steps:

- Install .NET 6 Preview 7
- `dotnet new blazorwasm -ho -o BlazorApp`
- `cd BlazorApp/Server`
- `dotnet watch`
- Update Index.razor to add the following simple binding code:

```
<input @bind="text" />
<p>@text</p>

@code {
  string? text;
}
```

- Wait for the app to restart
- Update the `Title` attribute of the `SurveyPrompt` component to change the attribute value to something else.

Expected result:

- The change is applied to the running app and shows up in the survey prompt in the browser

Actual result:

- `dotnet watch` says the change was successfully applied, but the change doesn't show up in the browser. Navigating to a different tab and back again to trigger a rerender doesn't help, neither does refreshing the browser.
- If you make another simple change, like updating the attribute value again or changing some HTML markup, `dotnet watch` says that a rude edit has occurred.

@pranavkm @lambdageek

**Comment by danroth27** [comment]
This issue also reproduces with a Blazor Server app, although subsequent changes don't trigger a rude edit like they do with Blazor WebAssembly. Changes just don't show up once the data binding code has been added even though `dotnet watch` says they were successfully applied.

**Comment by mkArtakMSFT** [comment]
@tmat here is another broken scenario, related to delegates changing in .razor files.

**Comment by pranavkm** [comment]
@lambdageek I was testing @tmat's changes to support enc without sequence points and that appears to be solved with the latest rc1 installer. Ordinary edits that do not update the delegate work fine and that seems great. 

Unfortunately, this particular scenario continues to fail. Here's the diff that we get:

ILDelta: 
```
AAAAABMwCAAdAQAAAAAAAAADFm8YAAArAAMXcp0OAHB+JgAABCUtFyZ+JQAABP4GRgAABnNSAAAKJYAmAAAEb1MAAAoAA29UAAAKAAMZcrcOAHBvVQAACgADGnLBDgBwb1UAAAoAAxtvGQAAKwADHHIxDwBwcj0PAHBvVgAACgADb1QAAAoAAx1yUQ8AcG9VAAAKAAMeclcPAHBvVwAACgADHwlyYw8AcAJ7CAAABBQoWAAACm9WAAAKAAMfCnJvDwBwflkAAAoCAv4GIAAABnNaAAAKAnsIAAAEFChbAAAKbxoAACsAA3JjDwBwb10AAAoAA29eAAAKAAMfC3JRDwBwb1UAAAoAAx8McoEPAHBvVwAACgADHw0CewgAAARvXwAACgADb14AAAoAKiICA30IAAAEKj4AAxhyhQ8AcG9fAAAKACo=
```

MetadataDelta:

```
QlNKQgEAAQAAAAAADAAAAHY0LjAuMzAzMTkAAAAABgB8AAAAGAUAACMtAACUBQAAvAIAACNTdHJpbmdzAAAAAFAIAAD4AAAAI1VTAEgJAAAwAAAAI0dVSUQAAAB4CQAApAAAACNCbG9iAAAAHAoAAAAAAAAjSlREAAAAAAAAAAACAKcBQxUAyAgIAAAA+gEzABYAAAEAAAAQAAAAAwAAAAMAAAARAAAAAwAAAAEAAAAxAAAAMQAAAAMAAAADAAAAAQCjDgAAAgAAAAMAAAAAAAAAKgAAAEgOAABlDwAAKgAAAAAOAABlDwAAKgAAANMOAABrDgAAJgAAAC0OAAAQDwAAJgAAAIUPAACrDgAALgAAAM8NAACbDQAAKgAAAKUPAABlDwAAJgAAAA4OAACrDgAAKgAAAPwOAABlDwAAJgAAAMcOAACyDgAAKgAAAJUOAABlDwAAKgAAAO0PAABlDwAAJgAAAIcNAACrDgAAKgAAAEAPAABlDwAAKgAAAHcNAABlDwAAKgAAADAPAABlDwAABAAAAAAAxAC/DQAAQgUAAAAAAAAtAQAAAACBAF8NAAC5BAAAAAAAADYBAAAAAIMARw0AAEIFAAAAAAAAAAABAOUOAAAAAAEAYw4AAAAAAQCQDQAAYQIAAAoPAAC5BAAAeQIAAAoPAAC+BAAAcQIAAMMPAADCBAAAkQIAAAoPAADOBAAAcQIAACAOAADUBAAAcQIAALQPAAC+BAAAcQIAANwPAADdBAAAcQIAACAOAADoBAAAcQIAAJkPAADdBAAAoQIAAFcOAADvBAAAsQIAAPoPAAD3BAAANAAAAAoPAADOBAAAyQIAAO8OAAADBQAAcQIAACAOAAAcBQAAcQIAANkNAAC5BAAAcQIAAIwPAAC+BAAAcQIAANEPAADdBAAAAAQAACIAAABJBQAAAAQAAIMCAABPBQAAwAgAACIAAABUBQAA/AQAAAkAACMAAAAACgAAIwAAAAALAAAjAAAAAE8AAAoAAAAAUAAACgAAAABRAAAKAAAAAFIAAAoAAAAAUwAACgAAAABUAAAKAAAAAFUAAAoAAAAAVgAACgAAAABXAAAKAAAAAFgAAAoAAAAAWQAACgAAAABaAAAKAAAAAFsAAAoAAAAAXAAACgAAAABdAAAKAAAAAF4AAAoAAAAAXwAACgAAAAAYAAArAAAAABkAACsAAAAAGgAAKwAAAABMAAABAAAAAE0AAAEAAAAATgAAAQAAAABPAAABAAAAAFAAAAEAAAAAUQAAAQAAAABSAAABAAAAAFMAAAEAAAAAVAAAAQAAAABVAAABAAAAAFYAAAEAAAAAVwAAAQAAAABYAAABAAAAAFkAAAEAAAAAWgAAAQAAAABbAAABAAAAAAYAABsAAAAAHgAABgAAAAAgAAAGAAAAAEYAAAYAAAAADAAACAAAAAANAAAIAAAAAB8AAAgAAAAAQgAADAAAAABDAAAMAAAAAFgAAAwAAAAATAAAAU0AAAFOAAABTwAAAVAAAAFRAAABUgAAAVMAAAFUAAABVQAAAVYAAAFXAAABWAAAAVkAAAFaAAABWwAAAR4AAAYgAAAGRgAABgwAAAgNAAAIHwAACE8AAApQAAAKUQAAClIAAApTAAAKVAAAClUAAApWAAAKVwAAClgAAApZAAAKWgAAClsAAApcAAAKXQAACl4AAApfAAAKQgAADEMAAAxYAAAMBgAAGwkAACMKAAAjCwAAIxgAACsZAAArGgAAKwYAAAAAAAAAAAAAADAFAADxDQAAAAAAAAAAAAAGAAAAAAAAAAAAAAA5BQAAZQ8AAAAAAAAAAAAABgAAAAAAAAAAAAAAOQUAAJsNAAAAAAAAAAAAAKMAAADIBAAAowAAAOMEAAC5AAAAKgUAAAAAADxCdWlsZFJlbmRlclRyZWU+Yl9fMF8wADxCdWlsZFJlbmRlclRyZWU+Yl9fMF8xAEV2ZW50Q2FsbGJhY2tgMQBBY3Rpb25gMQBfX2J1aWxkZXIyAE1pY3Jvc29mdC5Bc3BOZXRDb3JlLkNvbXBvbmVudHMuV2ViAEJ1aWxkUmVuZGVyVHJlZQBQYWdlVGl0bGUAU2V0VXBkYXRlc0F0dHJpYnV0ZU5hbWUAU3lzdGVtLlJ1bnRpbWUAQ29tcG9uZW50QmFzZQBNdWx0aWNhc3REZWxlZ2F0ZQBBZGRBdHRyaWJ1dGUAQ29tcGlsZXJHZW5lcmF0ZWRBdHRyaWJ1dGUAUm91dGVBdHRyaWJ1dGUARm9ybWF0VmFsdWUAX192YWx1ZQBNaWNyb3NvZnQuQXNwTmV0Q29yZS5Db21wb25lbnRzLlJlbmRlcmluZwBFdmVudENhbGxiYWNrAGJsMS5kbGwAU3lzdGVtAFN5c3RlbS5HbG9iYWxpemF0aW9uAEN1bHR1cmVJbmZvAFJlbmRlclRyZWVCdWlsZGVyAF9fYnVpbGRlcgBDcmVhdGVCaW5kZXIAQmluZENvbnZlcnRlcgAuY3RvcgBTeXN0ZW0uUnVudGltZS5Db21waWxlclNlcnZpY2VzAENoYW5nZUV2ZW50QXJncwBFdmVudENhbGxiYWNrRmFjdG9yeUJpbmRlckV4dGVuc2lvbnMATWljcm9zb2Z0LkFzcE5ldENvcmUuQ29tcG9uZW50cwBPYmplY3QAQ2xvc2VFbGVtZW50AE9wZW5FbGVtZW50AFJlbmRlckZyYWdtZW50AENsb3NlQ29tcG9uZW50AE9wZW5Db21wb25lbnQAQWRkQ29udGVudABBZGRNYXJrdXBDb250ZW50AEV2ZW50Q2FsbGJhY2tGYWN0b3J5AAAZQwBoAGkAbABkAEMAbwBuAHQAZQBuAHQAAAkNAAoADQAKAABvPABoADEAPgBIAGUAbABsAG8ALAAgAHcAbwByAGwAZAAhACEAPAAvAGgAMQA+AA0ACgANAAoAVwBlAGwAYwBvAG0AZQAgAHQAbwAgAHkAbwB1AHIAIABuAGUAdwAgAGEAcABwAC4ADQAKAA0ACgAAC1QAaQB0AGwAZQAAE0gAZQBsAGwAbwA/AD8APwA/AAAFDQAKAAALaQBuAHAAdQB0AAALdgBhAGwAdQBlAAARbwBuAGMAaABhAG4AZwBlAAADcAAAC0kAbgBkAGUAeAAAAAAAAAAAAAAAAAAAAAAAAAAAAH0ZNd2lrVBNqQzpVT625p0nEVxdxgXGRatZ6RPCMiMNAAQgAQEOAyAAAQUwAQEBCAUKARKBRQUgAgEcGAggAwEIDhKBTQUgAgEIDgQKARIgBiADAQgODgcAAg4OEoFVBAYSgV0GFRKBYQEOGAAFFRGBaQESgW0SgV0cFRKBYQEODhKBVQ0wAQMBCA4VEYFpAR4ABQoBEoFtCLA/X38R1Qo6CK25eTgp3a5gBiABARKBOQUBAAIAAAQBAAAABQEAAQAAAAA=
```

Are you able to help?

**Comment by lambdageek** [comment]
> Are you able to help?

Well I can at least reproduce it.  the delta looks like something that should apply cleanly.
I'm not sure yet:
1. Why the  page doesn't refresh when you change the SurveyPrompt Title attribute.
2. Why applying a second change causes an assertion.

**Comment by lambdageek** [comment]
It's something to do with `@bind` making an instance (`this`-capturing) lambda; in the case it's `__value => text = __value` where `text` is a field on the `Index` class.  That's the main difference from the RenderFragment lambdas which were all static.  Not sure yet why it makes a difference.  Although it also seems to affect `this`-capturing lambdas in `@code` sections.  So it's not the `#line` pragmas this time.

```
<button class="btn btn-primary" @onclick="IncrementCount">Click me</button>

@code {
    private int currentCount = 0;

    private void IncrementCount()
    {
        currentCount+=2;

        Action complicateIt = () => currentCount--;
        complicateIt();
    }
}
```
has the same weird behavior.

**Comment by lambdageek** [comment]
TL;DR:

1. There's a runtime bug where I need to be more lax about CustomAttribute updates. I'm going to open a runtime issue.
2. the `dotnet watch` delta applier catches exceptions and logs them, but doesn't actually consider that to cause a rude edit, AFAICT.

---

Ah... no, it's not quite the same kind of delta.

The lambda spits out an updated `[CompilerGenerated]` attribute on the method for the lambda and the runtime throws a `TypeLoadException` similar to this:
```
System.TypeLoadException: EnC: we do not support patching of existing CA table cols with a different Parent or Type. token=0x0c00000f
   at System.Reflection.Metadata.MetadataUpdater.ApplyUpdate(Assembly assembly, ReadOnlySpan`1 metadataDelta, ReadOnlySpan`1 ilDelta, ReadOnlySpan`1 pdbDelta)
   at MonoDelta.DeltaHelper.LoadMetadataUpdate(Assembly assm, Byte[] dmeta_data, Byte[] dil_data, Byte[] dpdb_data) in /Users/alklig/work/dotnet-runtime/runtime/src/mono/sample/mbr/DeltaHelper/DeltaHelper.cs:line 12
   at MonoDelta.DeltaHelper.Update(Assembly assm) in /Users/alklig/work/dotnet-runtime/runtime/src/mono/sample/mbr/DeltaHelper/DeltaHelper.cs:line 62
   at HelloWorld.Program.Main(String[] args) in /Users/alklig/work/dotnet-runtime/runtime/src/mono/sample/mbr/console/Program.cs:line 71
```

on the first (apparently successful edit).

Seems like the hot reload delta applier is catching the error (and logging it... somewhere...) but doesn't actually indicate that the delta failed to apply. 

https://github.com/dotnet/sdk/blob/baecd78dd60fc32f2cc942d3df6a520c28ac96ad/src/BuiltInTools/DotNetDeltaApplier/HotReloadAgent.cs#L212-L215

The assertion when the second delta is applied happens because the runtime rolls back the first bad delta when it throws the TypeLoadException, and it doesn't expect another delta to come in.

**Comment by lambdageek** [comment]
> This issue also reproduces with a Blazor Server app, although subsequent changes don't trigger a rude edit like they do with Blazor WebAssembly. Changes just don't show up once the data binding code has been added even though `dotnet watch` says they were successfully applied.

Also @pranavkm  I'm not sure what to make of this - these kinds of deltas are well within coreclr's capabilities - appllying them server-side should work.

**Comment by lambdageek** [comment]
The Mono issue will be fixed by https://github.com/dotnet/runtime/pull/57388

**Comment by lambdageek** [comment]
Verified that `dotnet watch` and blazorwasm and `@bind` work in 6.0.100-rc.1.21416.3

**Comment by pranavkm** [comment]
I'll update the code to throw. Probably better for the app to fail than have it try to log the exception.
